modules/objdetect/src/qrcode.cpp

```diff
@@ -2378,45 +2379,33 @@ bool QRDecode::samplingForVersion()
 
     const int delta_rows = cvRound((postIntermediate.rows * 1.0) / version_size);
     const int delta_cols = cvRound((postIntermediate.cols * 1.0) / version_size);
+    const int skipped_rows = postIntermediate.rows - delta_rows * version_size;
+    const int skipped_cols = postIntermediate.cols - delta_cols * version_size;
 
-    vector<double> listFrequencyElem;
-    for (int r = 0; r < postIntermediate.rows; r += delta_rows)
-    {
-        for (int c = 0; c < postIntermediate.cols; c += delta_cols)
-        {
+    vector<int> deltas_rows(version_size, delta_rows);
+    vector<int> deltas_cols(version_size, delta_cols);
+
+    for (int i = 0; i < abs(skipped_rows); i++) {
+        deltas_rows[(i*version_size)/abs(skipped_rows)+(version_size/abs(skipped_cols)-1)/2]
+        += skipped_rows > 0 ? 1 : -1;
+    }
+    for (int i = 0; i < abs(skipped_cols); i++) {
+        deltas_cols[(i*version_size)/abs(skipped_cols)+(version_size/abs(skipped_cols)-1)/2]
+        += skipped_cols > 0 ? 1 : -1;
+    }
+
+    const double totalFrequencyElem = countNonZero(postIntermediate) / static_cast<double>(postIntermediate.total());
```

Comment:
Looks like `postIntermediate` is not a binary (0/255) image. So there is no sense to use `countNonZero()` here.

Perhaps we want to use `norm(postIntermediate, NORM_L1 | NORM_RELATIVE);` here.

---

Do we have 0.5 value here?

Reply:
actually `postIntermediate` is binary (0/255) image in this case:

1. `no_border_intermediate` is binary
2. `postIntermediate` = `resize(no_border_intermediate, newFactorSize, INTER_AREA)`
3. `newFactorSize` is integer and  InterpolationFlags is `INTER_AREA` so `postIntermediate` is binary (0/255) image